File: internal/api/instances.go
package api

import (
	"errors"
	"net/http"
	"sort"
	"time"

	"github.com/abbyfluoroethane/bento/internal/store"
	"github.com/abbyfluoroethane/bento/internal/types"
)

// instanceJSON is the wire form of an instance row. Times are RFC 3339.
type instanceJSON struct {
	UUID         string `json:"uuid"`
	Name         string `json:"name"`
	Owner        string `json:"owner"`
	State        string `json:"state"`
	DesiredState string `json:"desired_state"`
	Address      string `json:"address"`
	MAC          string `json:"mac"`
	Image        string `json:"image"`
	BaseChecksum string `json:"base_checksum"`
	VCPU         int    `json:"vcpu"`
	MemoryMiB    int64  `json:"memory_mib"`
	DiskGiB      int64  `json:"disk_gib"`
	Nested       bool   `json:"nested"`
	KSM          bool   `json:"ksm"`
	HTTPPort     int    `json:"http_port"`
	Visibility   string `json:"visibility"`
	CreatedAt    string `json:"created_at"`
	LastSeenAt   string `json:"last_seen_at"`
	SharedWithMe bool   `json:"shared_with_me"`
}

type quotaJSON struct {
	MaxInstances int   `json:"max_instances"`
	MaxVCPU      int   `json:"max_vcpu"`
	MaxMemoryMiB int64 `json:"max_memory_mib"`
	MaxDiskGiB   int64 `json:"max_disk_gib"`
}

type usageJSON struct {
	Instances int   `json:"instances"`
	VCPU      int64 `json:"vcpu"`
	MemoryMiB int64 `json:"memory_mib"`
	DiskGiB   int64 `json:"disk_gib"`
}

func toUsageJSON(u store.Usage) usageJSON {
	return usageJSON{Instances: u.Instances, VCPU: u.VCPU, MemoryMiB: u.MemoryMiB, DiskGiB: u.DiskGiB}
}

func rfc3339(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.UTC().Format(time.RFC3339)
}

func (s *Server) toInstanceJSON(inst types.Instance, viewer types.User, owners map[int64]string) instanceJSON {
	owner, ok := owners[inst.OwnerID]
	if !ok {
		if u, err := s.cfg.Store.UserByID(inst.OwnerID); err == nil {
			owner = u.Name
		}
		owners[inst.OwnerID] = owner
	}
	return instanceJSON{
		UUID:         inst.UUID,
		Name:         inst.Name,
		Owner:        owner,
		State:        string(inst.State),
		DesiredState: string(inst.DesiredState),
		Address:      inst.Address,
		MAC:          inst.MAC,
		Image:        inst.ImageName,
		BaseChecksum: inst.BaseChecksum,
		VCPU:         inst.VCPU,
		MemoryMiB:    inst.MemoryMiB,
		DiskGiB:      inst.DiskGiB,
		Nested:       inst.Nested,
		KSM:          inst.KSM,
		HTTPPort:     inst.HTTPPort,
		Visibility:   string(inst.Visibility),
		CreatedAt:    rfc3339(inst.CreatedAt),
		LastSeenAt:   rfc3339(inst.LastSeenAt),
		SharedWithMe: inst.OwnerID != viewer.ID,
	}
}

type instanceListResponse struct {
	Instances []instanceJSON `json:"instances"`
	Quota     *quotaJSON     `json:"quota"`
	Usage     usageJSON      `json:"usage"`
}

// handleListInstances answers the primary dashboard view (SPEC 14.4): the
// user's instances plus the ones shared with them, sorted by name, with
// the four quota limits and current use for the bar above the table.
func (s *Server) handleListInstances(w http.ResponseWriter, r *http.Request, u types.User) {
	owned, err := s.cfg.Store.InstancesByOwner(u.ID)
	if err != nil {
		s.writeStoreError(w, err)
		return
	}
	shared, err := s.cfg.Store.InstancesSharedWith(u.ID)
	if err != nil {
		s.writeStoreError(w, err)
		return
	}
	usage, err := s.cfg.Store.UsageFor(u.ID)
	if err != nil {
		s.writeStoreError(w, err)
		return
	}

	resp := instanceListResponse{Instances: []instanceJSON{}, Usage: toUsageJSON(usage)}
	owners := map[int64]string{u.ID: u.Name}
	for _, inst := range append(owned, shared...) {
		resp.Instances = append(resp.Instances, s.toInstanceJSON(inst, u, owners))
	}
	sort.Slice(resp.Instances, func(i, j int) bool {
		return resp.Instances[i].Name < resp.Instances[j].Name
	})

	quota, err := s.cfg.Store.QuotaFor(u.ID)
	switch {
	case err == nil:
		resp.Quota = &quotaJSON{
			MaxInstances: quota.MaxInstances,
			MaxVCPU:      quota.MaxVCPU,
			MaxMemoryMiB: quota.MaxMemoryMiB,
			MaxDiskGiB:   quota.MaxDiskGiB,
		}
	case errors.Is(err, store.ErrNotFound):
		// No quotas row means unlimited; the dashboard shows use only.
	default:
		s.writeStoreError(w, err)
		return
	}

	writeJSON(w, http.StatusOK, resp)
}

func (s *Server) handleGetInstance(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, err := s.cfg.Store.Instance(r.PathValue("uuid"))
	if err != nil {
		s.writeStoreError(w, err)
		return
	}
	if inst.OwnerID != u.ID {
		shares, err := s.cfg.Store.SharesFor(inst.UUID)
		if err != nil || !sharedWith(shares, u.ID) {
			writeError(w, http.StatusNotFound, "not found")
			return
		}
	}
	owners := map[int64]string{u.ID: u.Name}
	writeJSON(w, http.StatusOK, s.toInstanceJSON(inst, u, owners))
}

type createRequest struct {
	Name      string `json:"name"`
	Image     string `json:"image"`
	VCPU      int    `json:"vcpu"`
	MemoryMiB int64  `json:"memory_mib"`
	DiskGiB   int64  `json:"disk_gib"`
	Nested    bool   `json:"nested"`
	// KSM defaults to on (SPEC 5.4); a pointer keeps "absent" distinct
	// from the JSON zero value false.
	KSM *bool `json:"ksm"`
}

func (s *Server) handleCreateInstance(w http.ResponseWriter, r *http.Request, u types.User) {
	var req createRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request body: "+err.Error())
		return
	}
	if !validName(req.Name) {
		writeError(w, http.StatusBadRequest,
			"instance name must be a DNS label: lower-case letters, digits, and hyphens, up to 63 characters")
		return
	}
	if req.Image == "" {
		writeError(w, http.StatusBadRequest, "image is required")
		return
	}
	if req.VCPU < 0 || req.MemoryMiB < 0 || req.DiskGiB < 0 {
		writeError(w, http.StatusBadRequest, "vcpu, memory_mib, and disk_gib must not be negative")
		return
	}
	spec := CreateSpec{
		Name:      req.Name,
		Image:     req.Image,
		VCPU:      req.VCPU,
		MemoryMiB: req.MemoryMiB,
		DiskGiB:   req.DiskGiB,
		Nested:    req.Nested,
		KSM:       req.KSM == nil || *req.KSM,
	}
	inst, err := s.cfg.Lifecycle.Create(r.Context(), u, spec)
	if err != nil {
		s.writeStoreError(w, err)
		return
	}
	owners := map[int64]string{u.ID: u.Name}
	writeJSON(w, http.StatusCreated, s.toInstanceJSON(inst, u, owners))
}

func (s *Server) handleDeleteInstance(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	if err := s.cfg.Lifecycle.Delete(r.Context(), inst.UUID); err != nil {
		s.writeStoreError(w, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// actionHandler builds the start/stop/restart handlers, which differ only
// in the lifecycle call.
func (s *Server) actionHandler(action string) func(http.ResponseWriter, *http.Request, types.User) {
	return func(w http.ResponseWriter, r *http.Request, u types.User) {
		inst, ok := s.ownedInstance(w, r, u)
		if !ok {
			return
		}
		var err error
		switch action {
		case "start":
			err = s.cfg.Lifecycle.Start(r.Context(), inst.UUID)
		case "stop":
			err = s.cfg.Lifecycle.Stop(r.Context(), inst.UUID)
		case "restart":
			err = s.cfg.Lifecycle.Restart(r.Context(), inst.UUID)
		}
		if err != nil {
			s.writeStoreError(w, err)
			return
		}
		writeJSON(w, http.StatusAccepted, map[string]string{"uuid": inst.UUID, "action": action})
	}
}

type renameRequest struct {
	NewName string `json:"new_name"`
}

// handleRename renames an instance. The confirmation dialog of SPEC 7.3
// (old links break, the SSH user name changes) is the dashboard's job;
// the API performs the rename it is asked for.
func (s *Server) handleRename(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	var req renameRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request body: "+err.Error())
		return
	}
	if !validName(req.NewName) {
		writeError(w, http.StatusBadRequest,
			"instance name must be a DNS label: lower-case letters, digits, and hyphens, up to 63 characters")
		return
	}
	if err := s.cfg.Lifecycle.Rename(r.Context(), inst.UUID, req.NewName); err != nil {
		s.writeStoreError(w, err)
		return
	}
	s.writeRefreshed(w, inst.UUID, u)
}

type resizeRequest struct {
	VCPU      int   `json:"vcpu"`
	MemoryMiB int64 `json:"memory_mib"`
	DiskGiB   int64 `json:"disk_gib"`
	// Nested is a pointer so "leave it alone" is distinct from false.
	Nested *bool `json:"nested"`
}

func (s *Server) handleResize(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	var req resizeRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request body: "+err.Error())
		return
	}
	spec := ResizeSpec{
		VCPU:      inst.VCPU,
		MemoryMiB: inst.MemoryMiB,
		DiskGiB:   inst.DiskGiB,
		Nested:    inst.Nested,
	}
	if req.VCPU != 0 {
		spec.VCPU = req.VCPU
	}
	if req.MemoryMiB != 0 {
		spec.MemoryMiB = req.MemoryMiB
	}
	if req.DiskGiB != 0 {
		spec.DiskGiB = req.DiskGiB
	}
	if req.Nested != nil {
		spec.Nested = *req.Nested
	}
	if spec.VCPU < 1 || spec.MemoryMiB < 1 {
		writeError(w, http.StatusBadRequest, "vcpu and memory_mib must be positive")
		return
	}
	// A qcow2 overlay only grows (SPEC 11.1); catching a shrink here gives
	// a clear message instead of a qemu-img failure.
	if spec.DiskGiB < inst.DiskGiB {
		writeError(w, http.StatusBadRequest, "the disk can grow but never shrink")
		return
	}
	if err := s.cfg.Lifecycle.Resize(r.Context(), inst.UUID, spec); err != nil {
		s.writeStoreError(w, err)
		return
	}
	s.writeRefreshed(w, inst.UUID, u)
}

type portRequest struct {
	Port int `json:"port"`
}

func (s *Server) handlePort(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	var req portRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request body: "+err.Error())
		return
	}
	if req.Port < 1 || req.Port > 65535 {
		writeError(w, http.StatusBadRequest, "port must be between 1 and 65535")
		return
	}
	if err := s.cfg.Lifecycle.SetHTTPPort(r.Context(), inst.UUID, req.Port); err != nil {
		s.writeStoreError(w, err)
		return
	}
	s.writeRefreshed(w, inst.UUID, u)
}

type visibilityRequest struct {
	Visibility string `json:"visibility"`
}

func (s *Server) handleVisibility(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	var req visibilityRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request body: "+err.Error())
		return
	}
	v := types.Visibility(req.Visibility)
	if v != types.VisibilityOff && v != types.VisibilityPrivate && v != types.VisibilityPublic {
		writeError(w, http.StatusBadRequest, `visibility must be "off", "private", or "public"`)
		return
	}
	// Through the lifecycle, not the store: a visibility change alters
	// the published ports, and SPEC 6.3 reloads the nftables table on
	// every change.
	if err := s.cfg.Lifecycle.SetVisibility(r.Context(), inst.UUID, v); err != nil {
		s.writeStoreError(w, err)
		return
	}
	s.writeRefreshed(w, inst.UUID, u)
}

// writeRefreshed answers a mutation with the instance as it now stands.
func (s *Server) writeRefreshed(w http.ResponseWriter, uuid string, u types.User) {
	inst, err := s.cfg.Store.Instance(uuid)
	if err != nil {
		s.writeStoreError(w, err)
		return
	}
	owners := map[int64]string{u.ID: u.Name}
	writeJSON(w, http.StatusOK, s.toInstanceJSON(inst, u, owners))
}

type shareJSON struct {
	User      string `json:"user"`
	CreatedAt string `json:"created_at"`
}

func (s *Server) handleListShares(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	shares, err := s.cfg.Store.SharesFor(inst.UUID)
	if err != nil {
		s.writeStoreError(w, err)
		return
	}
	out := []shareJSON{}
	for _, sh := range shares {
		name := ""
		if user, err := s.cfg.Store.UserByID(sh.UserID); err == nil {
			name = user.Name
		}
		out = append(out, shareJSON{User: name, CreatedAt: rfc3339(sh.CreatedAt)})
	}
	writeJSON(w, http.StatusOK, out)
}

type shareRequest struct {
	User string `json:"user"`
}

func (s *Server) handleAddShare(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	var req shareRequest
	if err := decodeJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request body: "+err.Error())
		return
	}
	target, err := s.cfg.Store.UserByName(req.User)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			writeError(w, http.StatusNotFound, "no user named "+req.User)
			return
		}
		s.writeStoreError(w, err)
		return
	}
	if target.ID == u.ID {
		writeError(w, http.StatusBadRequest, "you already own this instance")
		return
	}
	if err := s.cfg.Store.AddShare(inst.UUID, target.ID); err != nil {
		s.writeStoreError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, shareJSON{User: target.Name})
}

func (s *Server) handleRemoveShare(w http.ResponseWriter, r *http.Request, u types.User) {
	inst, ok := s.ownedInstance(w, r, u)
	if !ok {
		return
	}
	target, err := s.cfg.Store.UserByName(r.PathValue("user"))
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			writeError(w, http.StatusNotFound, "no user named "+r.PathValue("user"))
			return
		}
		s.writeStoreError(w, err)
		return
	}
	if err := s.cfg.Store.RemoveShare(inst.UUID, target.ID); err != nil {
		s.writeStoreError(w, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

File: crates/api/src/instances.rs
use axum::body::Body;
use axum::extract::{Extension, Path as AxumPath, State};
use axum::http::{Request, StatusCode};
use axum::response::{IntoResponse, Response};
use bento_store::Usage;
use bento_types::{Instance, Quota, User, Visibility};
use serde::{Deserialize, Serialize};

use crate::{
    AppState, BAD_NAME, CreateSpec, ResizeSpec, decode_json, error_response, is_not_found,
    json_response, mapped_error, optional_rfc3339, rfc3339, valid_name,
};

#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct InstanceJson {
    pub(crate) uuid: String,
    pub(crate) name: String,
    pub(crate) owner: String,
    pub(crate) state: String,
    pub(crate) desired_state: String,
    pub(crate) address: String,
    pub(crate) mac: String,
    pub(crate) image: String,
    pub(crate) base_checksum: String,
    pub(crate) vcpu: u32,
    pub(crate) memory_mib: i64,
    pub(crate) disk_gib: i64,
    pub(crate) nested: bool,
    pub(crate) ksm: bool,
    pub(crate) http_port: u16,
    pub(crate) visibility: String,
    pub(crate) created_at: String,
    pub(crate) last_seen_at: String,
    pub(crate) shared_with_me: bool,
}

pub(crate) async fn instance_json(
    state: &AppState,
    instance: Instance,
    viewer: &User,
    owners: &mut std::collections::HashMap<i64, String>,
) -> InstanceJson {
    let owner = if let Some(owner) = owners.get(&instance.owner_id) {
        owner.clone()
    } else {
        let owner = state
            .0
            .store
            .user_by_id(instance.owner_id)
            .await
            .map(|user| user.name)
            .unwrap_or_default();
        owners.insert(instance.owner_id, owner.clone());
        owner
    };
    InstanceJson {
        uuid: instance.uuid,
        name: instance.name,
        owner,
        state: instance.state.to_string(),
        desired_state: instance.desired_state.to_string(),
        address: instance.address,
        mac: instance.mac,
        image: instance.image_name,
        base_checksum: instance.base_checksum,
        vcpu: instance.vcpu,
        memory_mib: instance.memory_mib,
        disk_gib: instance.disk_gib,
        nested: instance.nested,
        ksm: instance.ksm,
        http_port: instance.http_port,
        visibility: instance.visibility.to_string(),
        created_at: rfc3339(instance.created_at),
        last_seen_at: optional_rfc3339(instance.last_seen_at),
        shared_with_me: instance.owner_id != viewer.id,
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct QuotaJson {
    pub(crate) max_instances: i64,
    pub(crate) max_vcpu: i64,
    pub(crate) max_memory_mib: i64,
    pub(crate) max_disk_gib: i64,
}

impl From<Quota> for QuotaJson {
    fn from(quota: Quota) -> Self {
        Self {
            max_instances: quota.max_instances,
            max_vcpu: quota.max_vcpu,
            max_memory_mib: quota.max_memory_mib,
            max_disk_gib: quota.max_disk_gib,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct UsageJson {
    pub(crate) instances: i64,
    pub(crate) vcpu: i64,
    pub(crate) memory_mib: i64,
    pub(crate) disk_gib: i64,
}

impl From<Usage> for UsageJson {
    fn from(usage: Usage) -> Self {
        Self {
            instances: usage.instances,
            vcpu: usage.vcpu,
            memory_mib: usage.memory_mib,
            disk_gib: usage.disk_gib,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct InstanceListResponse {
    pub(crate) instances: Vec<InstanceJson>,
    pub(crate) quota: Option<QuotaJson>,
    pub(crate) usage: UsageJson,
}

/// Answers the primary dashboard view (SPEC 14.4): owned and shared
/// instances sorted by name, plus all four quota limits and current use.
pub(crate) async fn list_instances(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
) -> Response {
    let owned = match state.0.store.instances_by_owner(user.id).await {
        Ok(instances) => instances,
        Err(error) => return mapped_error(error),
    };
    let shared = match state.0.store.instances_shared_with(user.id).await {
        Ok(instances) => instances,
        Err(error) => return mapped_error(error),
    };
    let usage = match state.0.store.usage_for(user.id).await {
        Ok(usage) => usage,
        Err(error) => return mapped_error(error),
    };
    let quota = match state.0.store.quota_for(user.id).await {
        Ok(quota) => Some(quota.into()),
        Err(error) if is_not_found(&error) => None,
        Err(error) => return mapped_error(error),
    };

    let mut owners = std::collections::HashMap::from([(user.id, user.name.clone())]);
    let mut instances = Vec::with_capacity(owned.len() + shared.len());
    for instance in owned.into_iter().chain(shared) {
        instances.push(instance_json(&state, instance, &user, &mut owners).await);
    }
    instances.sort_by(|left, right| left.name.cmp(&right.name));
    json_response(
        StatusCode::OK,
        &InstanceListResponse {
            instances,
            quota,
            usage: usage.into(),
        },
    )
}

pub(crate) async fn can_read_instance(state: &AppState, instance: &Instance, user: &User) -> bool {
    if instance.owner_id == user.id {
        return true;
    }
    state
        .0
        .store
        .shares_for(&instance.uuid)
        .await
        .is_ok_and(|shares| shares.iter().any(|share| share.user_id == user.id))
}

pub(crate) async fn get_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
) -> Response {
    let instance = match state.0.store.instance(&uuid).await {
        Ok(instance) => instance,
        Err(error) => return mapped_error(error),
    };
    if !can_read_instance(&state, &instance, &user).await {
        return error_response(StatusCode::NOT_FOUND, "not found");
    }
    let mut owners = std::collections::HashMap::from([(user.id, user.name.clone())]);
    json_response(
        StatusCode::OK,
        &instance_json(&state, instance, &user, &mut owners).await,
    )
}

pub(crate) async fn owned_instance(
    state: &AppState,
    uuid: &str,
    user: &User,
) -> Result<Instance, Box<Response>> {
    let instance = state
        .0
        .store
        .instance(uuid)
        .await
        .map_err(|error| Box::new(mapped_error(error)))?;
    if instance.owner_id == user.id {
        return Ok(instance);
    }
    match state.0.store.shares_for(&instance.uuid).await {
        Ok(shares) if shares.iter().any(|share| share.user_id == user.id) => Err(Box::new(
            error_response(StatusCode::FORBIDDEN, "only the owner may do this"),
        )),
        _ => Err(Box::new(error_response(StatusCode::NOT_FOUND, "not found"))),
    }
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct CreateRequest {
    name: String,
    #[serde(default)]
    image: String,
    #[serde(default)]
    vcpu: i64,
    #[serde(default)]
    memory_mib: i64,
    #[serde(default)]
    disk_gib: i64,
    #[serde(default)]
    nested: bool,
    ksm: Option<bool>,
}

pub(crate) async fn create_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    request: Request<Body>,
) -> Response {
    let request: CreateRequest = match decode_json(request).await {
        Ok(request) => request,
        Err(response) => return *response,
    };
    if !valid_name(&request.name) {
        return error_response(StatusCode::BAD_REQUEST, BAD_NAME);
    }
    if request.image.is_empty() {
        return error_response(StatusCode::BAD_REQUEST, "image is required");
    }
    if request.vcpu < 0 || request.memory_mib < 0 || request.disk_gib < 0 {
        return error_response(
            StatusCode::BAD_REQUEST,
            "vcpu, memory_mib, and disk_gib must not be negative",
        );
    }
    let Ok(vcpu) = u32::try_from(request.vcpu) else {
        return error_response(StatusCode::BAD_REQUEST, "vcpu is too large");
    };
    let spec = CreateSpec {
        name: request.name,
        image: request.image,
        vcpu,
        memory_mib: request.memory_mib,
        disk_gib: request.disk_gib,
        nested: request.nested,
        ksm: request.ksm.unwrap_or(true),
    };
    let instance = match state.0.lifecycle.create(user.clone(), spec).await {
        Ok(instance) => instance,
        Err(error) => return mapped_error(error),
    };
    let mut owners = std::collections::HashMap::from([(user.id, user.name.clone())]);
    json_response(
        StatusCode::CREATED,
        &instance_json(&state, instance, &user, &mut owners).await,
    )
}

pub(crate) async fn delete_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
) -> Response {
    let instance = match owned_instance(&state, &uuid, &user).await {
        Ok(instance) => instance,
        Err(response) => return *response,
    };
    match state.0.lifecycle.delete(&instance.uuid).await {
        Ok(()) => StatusCode::NO_CONTENT.into_response(),
        Err(error) => mapped_error(error),
    }
}

#[derive(Serialize)]
pub(crate) struct ActionResponse<'a> {
    action: &'a str,
    uuid: &'a str,
}

pub(crate) async fn lifecycle_action(
    state: AppState,
    user: User,
    uuid: String,
    action: &'static str,
) -> Response {
    let instance = match owned_instance(&state, &uuid, &user).await {
        Ok(instance) => instance,
        Err(response) => return *response,
    };
    let result = match action {
        "start" => state.0.lifecycle.start(&instance.uuid).await,
        "stop" => state.0.lifecycle.stop(&instance.uuid).await,
        "restart" => state.0.lifecycle.restart(&instance.uuid).await,
        _ => unreachable!(),
    };
    match result {
        Ok(()) => json_response(
            StatusCode::ACCEPTED,
            &ActionResponse {
                action,
                uuid: &instance.uuid,
            },
        ),
        Err(error) => mapped_error(error),
    }
}

pub(crate) async fn start_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
) -> Response {
    lifecycle_action(state, user, uuid, "start").await
}

pub(crate) async fn stop_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
) -> Response {
    lifecycle_action(state, user, uuid, "stop").await
}

pub(crate) async fn restart_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
) -> Response {
    lifecycle_action(state, user, uuid, "restart").await
}

pub(crate) async fn refreshed(state: &AppState, uuid: &str, user: &User) -> Response {
    let instance = match state.0.store.instance(uuid).await {
        Ok(instance) => instance,
        Err(error) => return mapped_error(error),
    };
    let mut owners = std::collections::HashMap::from([(user.id, user.name.clone())]);
    json_response(
        StatusCode::OK,
        &instance_json(state, instance, user, &mut owners).await,
    )
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct RenameRequest {
    new_name: String,
}

/// The confirmation dialog (old links break and the SSH user changes) is the
/// dashboard's job; this endpoint performs only the explicit rename requested.
pub(crate) async fn rename_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
    request: Request<Body>,
) -> Response {
    let instance = match owned_instance(&state, &uuid, &user).await {
        Ok(instance) => instance,
        Err(response) => return *response,
    };
    let request: RenameRequest = match decode_json(request).await {
        Ok(request) => request,
        Err(response) => return *response,
    };
    if !valid_name(&request.new_name) {
        return error_response(StatusCode::BAD_REQUEST, BAD_NAME);
    }
    match state
        .0
        .lifecycle
        .rename(&instance.uuid, &request.new_name)
        .await
    {
        Ok(()) => refreshed(&state, &instance.uuid, &user).await,
        Err(error) => mapped_error(error),
    }
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct ResizeRequest {
    #[serde(default)]
    vcpu: i64,
    #[serde(default)]
    memory_mib: i64,
    #[serde(default)]
    disk_gib: i64,
    nested: Option<bool>,
}

pub(crate) async fn resize_instance(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
    request: Request<Body>,
) -> Response {
    let instance = match owned_instance(&state, &uuid, &user).await {
        Ok(instance) => instance,
        Err(response) => return *response,
    };
    let request: ResizeRequest = match decode_json(request).await {
        Ok(request) => request,
        Err(response) => return *response,
    };
    let vcpu = if request.vcpu == 0 {
        instance.vcpu
    } else if let Ok(vcpu) = u32::try_from(request.vcpu) {
        vcpu
    } else {
        return error_response(
            StatusCode::BAD_REQUEST,
            "vcpu and memory_mib must be positive",
        );
    };
    let spec = ResizeSpec {
        vcpu,
        memory_mib: if request.memory_mib == 0 {
            instance.memory_mib
        } else {
            request.memory_mib
        },
        disk_gib: if request.disk_gib == 0 {
            instance.disk_gib
        } else {
            request.disk_gib
        },
        nested: request.nested.unwrap_or(instance.nested),
    };
    if spec.vcpu < 1 || spec.memory_mib < 1 {
        return error_response(
            StatusCode::BAD_REQUEST,
            "vcpu and memory_mib must be positive",
        );
    }
    // A qcow2 overlay only grows (SPEC 11.1); catching a shrink here gives a
    // clear message instead of a qemu-img failure.
    if spec.disk_gib < instance.disk_gib {
        return error_response(
            StatusCode::BAD_REQUEST,
            "the disk can grow but never shrink",
        );
    }
    match state.0.lifecycle.resize(&instance.uuid, spec).await {
        Ok(()) => refreshed(&state, &instance.uuid, &user).await,
        Err(error) => mapped_error(error),
    }
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct PortRequest {
    port: i64,
}

pub(crate) async fn set_port(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
    request: Request<Body>,
) -> Response {
    let instance = match owned_instance(&state, &uuid, &user).await {
        Ok(instance) => instance,
        Err(response) => return *response,
    };
    let request: PortRequest = match decode_json(request).await {
        Ok(request) => request,
        Err(response) => return *response,
    };
    let Ok(port) = u16::try_from(request.port) else {
        return error_response(StatusCode::BAD_REQUEST, "port must be between 1 and 65535");
    };
    if port == 0 {
        return error_response(StatusCode::BAD_REQUEST, "port must be between 1 and 65535");
    }
    match state.0.lifecycle.set_http_port(&instance.uuid, port).await {
        Ok(()) => refreshed(&state, &instance.uuid, &user).await,
        Err(error) => mapped_error(error),
    }
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct VisibilityRequest {
    visibility: String,
}

pub(crate) async fn set_visibility(
    State(state): State<AppState>,
    Extension(user): Extension<User>,
    AxumPath(uuid): AxumPath<String>,
    request: Request<Body>,
) -> Response {
    let instance = match owned_instance(&state, &uuid, &user).await {
        Ok(instance) => instance,
        Err(response) => return *response,
    };
    let request: VisibilityRequest = match decode_json(request).await {
        Ok(request) => request,
        Err(response) => return *response,
    };
    let visibility = match request.visibility.as_str() {
        "off" => Visibility::Off,
        "private" => Visibility::Private,
        "public" => Visibility::Public,
        _ => {
            return error_response(
                StatusCode::BAD_REQUEST,
                "visibility must be \"off\", \"private\", or \"public\"",
            );
        }
    };
    // Through the lifecycle, not the store: visibility alters published
    // ports, and SPEC 6.3 reloads the nftables table on every change.
    match state
        .0
        .lifecycle
        .set_visibility(&instance.uuid, visibility)
        .await
    {
        Ok(()) => refreshed(&state, &instance.uuid, &user).await,
        Err(error) => mapped_error(error),
    }
}
